#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 opened the pull request:
- Added possibility to move more than one data point by using checkboxes
 - Added possibility to delete more than one folder / move more than one data point to root folder

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

 - Added possibility to move more than one data point by using checkboxes
 - Added possibility to delete more than one folder / move more than one data point to root folder

fabiodurao commented:
Thank you for paying attention to this topic, it was really a complex task to move one datapoint at a time in more complex projects. Both the traditional UI and the new UI suffered from this. I usually structure my projects by naming the datapoints with standardized xid's (e.g. UN1_voltage1, UN1_voltage2) which made it easier to move them in batches using json if I previously wrote them down in a spreadsheet, but if they are not written down, it becomes complicated because there is a problem in json that may require some discussion and the opening of a feature request, if you think it is better. The problem with json is that datapoints that are not yet in a folder do not appear in the json export, which makes it difficult to copy and paste them into a specific folder using json. In my opinion, it would be interesting if json also showed the xid's of datapoints that are in the root.

Limraj commented:
Hi @fabiodurao,
Did I understand correctly, do you think that points that are not in folders are not exported to json? I am asking you once again to write only about things that you are 100% sure about, i.e. which you checked and tested just before writing the comment. If you managed to repeat this problem now, please write what version of Scada-LTS you are using and the specific steps that led you to this error. 
From the description you have provided here, I cannot repeat it, the points are exported to a json file and I do it often and I have never had a point not exported, regardless of whether the point is in a folder or not. Moreover, the structure of folders with points is also exported. The export of points is independent from the export of folders, these are actually completely different structures.

Regards,
Kamil Jarmusik

fabiodurao commented:
Hello Kamil,

I apologize if the previous message was not clear. In the context of the previous message, I referred to the hierarchy, when it is exported, it is only possible to see the datapoints that are in a certain folder, as can be seen in the following images. I am testing this in SLTS 2.7.8. If I export the datapoints, they all appear, but it is a complex task to extract only the xid in the json of the datapoints, so I suggested that the json export of the hierarchy could show all the xid's that exist in a "root folder" for example, making it easier to move them to another folder by manipulating only the json.

![image](https://github.com/user-attachments/assets/556b18bb-c784-4aab-9132-852db3cc27aa)

![image](https://github.com/user-attachments/assets/01017c8c-a3f5-4251-a0ac-c4019c30989d)

Limraj commented:
Hi @fabiodurao,
This is an export of the hierarchy of this folder structure not of all points, that is, if you import this with all points (array dataPoints), you will get exactly this structure. In hierarchy, there is no details about points, only their xid and information about the folder they are in, if they are not in any, it does not apply to this hierarchy. So this is sufficient, adds no value. Please, Fabio, don't create unnecessary problems.

Regards,
Kamil Jarmusik

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Limraj commented on src/org/scada_lts/web/mvc/api/PointHierarchyAPI.java:
```diff
@@ -329,4 +327,71 @@ public ResponseEntity<PointHierarchyNode> getPointHierarchyWithEmptyRoot(HttpSer
         }
     }
+
+    @RequestMapping(value = "/api/pointHierarchy/moveBatch", method = RequestMethod.POST)
```
old: 
PUT /api/pointHierarchy/pointMoveTo/{xid_point}/{xid_folder}
then:
PUT /api/pointHierarchy/pointsMoveTo/{xid_folder}?xids=DP_XID_1;DP_XID_2&children=XID_1;XID_2

old:
PUT /api/pointHierarchy/folderMoveTo/{xid_folder}/{new_parent_xid_folder}
then:
PUT /api/pointHierarchy/foldersMoveTo/{xid_folder}?xids=FOLDER_XID_1;FOLDER_XID_2&children=XID_1;XID_2

Limraj commented on src/org/scada_lts/web/mvc/api/PointHierarchyAPI.java:
```diff
@@ -329,4 +327,71 @@ public ResponseEntity<PointHierarchyNode> getPointHierarchyWithEmptyRoot(HttpSer
+    }
+
+    @RequestMapping(value = "/api/pointHierarchy/deleteFolderMoveToRootBatch", method = RequestMethod.POST)
+    public ResponseEntity<String> deleteFolderMoveToRootBatch(
```
Isn't the old endpoint enough?
"/api/pointHierarchy/deleteFolder/{xid_folder}"

After all, the deletion mechanism should work as before, the change only concerned moving multiple points at once?

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Patrykb0802 committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.0' into feature/#2760_Improve_point_hierarchy_usage_move_several_datapoints_simultaneously

Limraj committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously:
- Added classes: DeleteObjectHierarchyDTO, MoveObjectHierarchyDTO, ObjectHierarchy, ObjectHierarchyType, CreatePointHierarchyTree;
- Corrected determination type object (POINT/DIR), instead of basing it on xid format, a ObjectHierarchyType class was created and the type determination is done on the frontend side;
- Refactoring classes: PointHierarchyAPI PointHierarchyXidService;
- Corrected class PointHierarchyUtils, create tree move to class CreatePointHierarchyTree, and execution in common ForkJoinPool;
- Corrected test: GetPointHierarchyRootTest with using common ForkJoinPool;

Limraj committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously:
- Fixed import/export point hierarchy on old UI;
- Fixed Point Hierarchy on new UI; (bad url)
- Refactored classes: PointHierarchyAPI, DeleteObjectHierarchyDTO, MoveObjectHierarchyDTO, ObjectHierarchy, ObjectHierarchyType, PointHierarchyXidService;

Limraj committed:
#2760 Improve point hierarchy usage : move several datapoints simultaneously:
- Corrected import css style in pointHierarchySLTS.jsp,
- Removed console.log in ExportImportPointHierarchy.vue,
- Corrected name method config in test GetPointHierarchyRootTest;
